sync/src/block/extension.rs

```diff
@@ -75,6 +75,64 @@ enum State {
     Full,
 }
 
+impl State {
+    fn initial(client: &Client, snapshot_target: Option<(BlockHash, u64)>) -> Self {
+        let (hash, num) = match snapshot_target {
+            Some(target) => target,
+            None => return State::Full,
+        };
+        let header = match client.block_header(&num.into()) {
+            Some(ref h) if h.hash() == hash => h.clone(),
```

Comment:
Cloning `header` only needed when the block body doesn't exist to return `State::SnapshotBody`, and it's an exceptional case.
I think it's better to remove unnecessary cloning because cloning header is an expensive job.

Reply:
```rust
        match client.block_header(&num.into()) {
            Some(ref header) if header.hash() == hash => {
                if client.block_body(&hash.into()).is_none() {
                    let parent_hash = header.parent_hash();
                    let parent =
                        client.block_header(&parent_hash.into()).expect("Parent header of the snapshot header must exist");
                    return State::SnapshotBody {
                        header: header.clone(),
                        prev_root: parent.transactions_root(),
                    }
                }

                let state_db = client.state_db().read();
                let state_root = header.state_root();
                match TrieFactory::readonly(state_db.as_hashdb(), &state_root) {
                    Ok(ref trie) if trie.is_complete() => State::Full,
                    _ => State::SnapshotChunk {
                        block: hash,
                        restore: SnapshotRestore::new(state_root),
                    },
                }
            },
            _ => State::SnapshotHeader(hash, num),
        }
```

Reply:
I think the performance is not an important thing to consider since this function is executed only once at the initialization.
I'd like to emphasize the readability of this part, and I tried to make the function check-and-return for each enum items.
What do you think in this perspective?

Reply:
I think the performance can be a matter even it's executed only at the initialization.
The slow startup sometimes makes a problem.
Moreover, I don't think the previous code is more readable.

Reply:
It suddenly come to mind that the if guard allows move since rust 1.39.0.
You can change it to
```rust
        let header = match client.block_header(&num.into()) {
            Some(h) if h.hash() == hash => h,
```

Reply:
Seems nice. I'll change it like that.